main/app-menu.js

```diff
@@ -0,0 +1,56 @@
+const { Menu, MenuItem, ipcMain } = require('electron')
+const { ipcMainEvents } = require('./ipc')
+
+function setupAppMenu (/** @type {import('./typings').Context} */ ctx) {
+  if (process.platform !== 'darwin') return
+
+  // Add "Check for updates..." item to the Application menu on MacOS
+  const menu = Menu.getApplicationMenu()
+  if (!menu) return
+
+  menu.items[0].submenu?.insert(1, new MenuItem({
+    id: 'checkForUpdates',
+    label: 'Check For Updates...',
+    click: () => { ctx.manualCheckForUpdates() }
+  }))
+  menu.items[0].submenu?.insert(2, new MenuItem({
+    id: 'checkingForUpdates',
+    label: 'Checking For Updates',
+    enabled: false,
+    visible: false
+  }))
+
+  Menu.setApplicationMenu(menu)
+  setupIpcEventListeners(menu)
+}
+
+/**
+ * @param {Electron.Menu} appMenu
+ */
+function setupIpcEventListeners (appMenu) {
+  ipcMain.on(ipcMainEvents.UPDATE_CHECK_STARTED, () => {
+    getItemById('checkForUpdates').enabled = false
+    getItemById('checkForUpdates').visible = false
+    getItemById('checkingForUpdates').visible = true
+  })
+
+  ipcMain.on(ipcMainEvents.UPDATE_CHECK_FINISHED, () => {
+    getItemById('checkForUpdates').enabled = true
+    getItemById('checkForUpdates').visible = true
+    getItemById('checkingForUpdates').visible = false
+  })
+
+  /**
+   * @param {string} id
+   * @returns {Electron.MenuItem}
+   */
+  function getItemById (id) {
+    const item = appMenu.getMenuItemById(id)
+    if (!item) throw new Error(`Unknown app menu item id: ${id}`)
+    return item
+  }
```

Comment:
Could we also do without this function? We'd get an error in the consuming lines anyway.

Reply:
Same question applies to the tray configuration

Reply:
This function serves two purposes:
- Make the TypeScript compiler happy
- Report a more helpful error message than _Uncaught TypeError: Cannot set properties of undefined (setting 'visible')`_

Reply:
Is this even run through ts, given the .js extension?

Reply:
Yes, we check types in JS files, that's why there are JSDoc comments in most places.

https://github.com/filecoin-project/filecoin-station/blob/62b5de84631c0bd3b01aa70407f4f1ccdf6ca021/tsconfig.json#L7-L8

I found this setup useful so far. I have full auto-completion support in VS Code and the type checker can catch many programming errors.